### PR TITLE
Check prisoner category before creating a move

### PR DESCRIPTION
**Pull request by solidgoldpig**
### What changed

After a user selects a prisoner to be moved, checks whether they should be moved or not.

If not, displays a page explaining that the person cannot be moved and why not.

### Why did it change

Prevents prisoners being inadvertently moved. 

### Issue tracking

- [P4-2310 - Check category at "Who is being moved step"](https://dsdmoj.atlassian.net/browse/P4-2310)

## Screenshots

<kbd>Error__ARCAMONE__AMERY_cannot_be_moved_using_this_service</kbd>

#### Automated testing

- [x] Added unit tests to cover changes
- [ ] Added end-to-end tests to cover any journey changes

#### Manual testing

Has been tested in the following browsers:

- [x] Chrome
- [ ] Firefox
- [ ] Edge
- [ ] Internet Explorer

### Environment variables

- [x] No environment variables were added or changed

### Other considerations

- [x] No new [Personally Identifiable Information (PII)](https://support.google.com/analytics/answer/6366371?hl=en) is being sent via analytics

### CODE REVIEW

File: app/move/views/create/move-not-supported.njk [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+{% block actionPreventedMessage %}
+  {{ t("validation::move_not_supported.message", {
+    category: category.key
+  }) | safe }}
+{% endblock %}
+
+{% block actionPreventedInstructions %}{% endblock %}
+
+{# {% block contentHeader %}
```
Comment (solidgoldpig) [inline app/move/views/create/move-not-supported.njk:25]: Remove the example bits Alex!!!

File: common/services/person.js [hunk 2]
```diff
@@ -56,6 +58,25 @@ const personService = {
...
+
+  getById(id) {
+    const include = personService.defaultInclude
+    return personService._getById(id, { include })
+  },
+
+  async getCategory(id) {
+    const include = ['category']
+    const person = await personService._getById(id, { include })
+    return person.category
```
Comment (teneightfive) [inline common/services/person.js:77]: This could also return a promise inline with the other service methods:

```javascript
getCategory(id) {
  return personService
    ._getById(id, { include: ['category'] })
    .then(person => person.category)
}
```

File: app/move/controllers/create/base.js [hunk 3]
```diff
@@ -42,6 +44,23 @@ class CreateBaseController extends FormWizardController {
     next()
   }
 
+  async checkMoveSupported(req, res, next) {
```
Comment (teneightfive) [inline app/move/controllers/create/base.js:47]: I've found these kind of methods easier to read when written using an early return pattern but that's purely personal preference:

```javascript
async checkMoveSupported(req, res, next) {
  const person = req.getPerson()

  if (!person || !person.prison_number) {
    return next()
  }

  const category = await personService.getCategory(person.id)
  if (!category || category.move_supported) {
    return next()
  }

  return res.render('move/views/create/move-not-supported', {
    person,
    category,
  })
}
```

The person check could even be simplified to (ie, if the person doesn't exist, by default they can't have a prison number?):

```javascript
  if (!person?.prison_number) {
    return next()
  }
```

**Review by teneightfive — APPROVED** [review]
Few minor style comments but looks good.

File: common/templates/layouts/action-prevented.njk [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+{% endblock %}
+
+{% block beforeContent %}
+  {% block actionPreventedBackLink %}{% endblock %}
+  {{ super() }}
+{% endblock %}
+
+{% block contentHeader %}
+  {% set title %}
+    {% block actionPreventedTitle %}{% endblock %}
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply common/templates/layouts/action-prevented.njk]: Good call. I'd misremembered that you couldn't set stuff outside of the blocks (except at load time). That's much, much cleaner...

File: app/move/views/create/move-not-supported.njk [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{% extends "layouts/action-prevented.njk" %}
+
+{% set actionPreventedTitle %}
```
Comment (teneightfive) [inline app/move/views/create/move-not-supported.njk:3]: A small note, but these can also be defined without the need for a block:

```nunjucks
{% set actionPreventedTitle = t("validation::move_not_supported.heading", {
  name: person.fullname
}) %}
```

The benefit of this is that you won't be injecting unnecessary whitespace into the final markup.

File: common/lib/api-client/models/index.js [hunk 6]
```diff
@@ -34,6 +35,7 @@ module.exports = {
   allocation_complex_case: allocationComplexCase,
   assessment_question: assessmentQuestion,
   cancel,
+  category,
```
Comment (teneightfive) [inline common/lib/api-client/models/index.js:38]: This might be a little too late to determine if the API side has already been built but `category` feels a little ambiguous. I wonder if it would be better named as `prison_category` or `prisoner_category` (I can't remember which the Prison API uses..)?

File: common/lib/api-client/models/person.js [hunk 7]
```diff
@@ -23,6 +23,10 @@ module.exports = {
       jsonApi: 'hasMany',
       type: 'profiles',
     },
+    category: {
```
Comment (teneightfive) [inline common/lib/api-client/models/person.js:26]: I'd also say, even if we don't want to rename the model. JSON:API allows the ability to define relationships with different names. So maybe at a minimum this relationship should be `prison_category` (or `prisoner_category` is that's the correct term).

File: common/lib/api-client/models/index.js (hunk 6 above)
Comment (solidgoldpig) [reply common/lib/api-client/models/index.js:38]: Will see what can be done

File: common/lib/api-client/models/person.js (hunk 7 above)
Comment (solidgoldpig) [reply common/lib/api-client/models/person.js:26]: > Will see what can be done

**Comment by jason-sims** [comment]
LGTM

File: common/lib/api-client/models/index.js (hunk 6 above)
Comment (solidgoldpig) [reply common/lib/api-client/models/index.js:38]: Apparently the use of `category` is intentional as locations have also been rejigged to have them.
